azure-eventhubs-eph/src/main/java/com/microsoft/azure/eventprocessorhost/PartitionManager.java

```diff
@@ -40,43 +43,68 @@
         this.host = host;
     }
     
-    String[] getPartitionIds() throws IllegalEntityException
+    CompletableFuture<String[]> getPartitionIds()
     {
-        Throwable saved = null;
-
-        if (this.partitionIds == null)
+    	CompletableFuture<String[]> retval = null;
+    	
+    	if (this.partitionIds != null)
+    	{
+    		retval = CompletableFuture.completedFuture(this.partitionIds);
+    	}
+    	else
         {
-			try
-			{
-				EventHubClient ehClient = EventHubClient.createFromConnectionStringSync(this.host.getEventHubConnectionString());
-				EventHubRuntimeInformation ehInfo = ehClient.getRuntimeInformation().get();
-				if (ehInfo != null)
+    		// This try-catch is necessary because EventHubClient.createFromConnectionString can directly throw
+    		// EventHubException or IOException, in addition to whatever failures may occur when the result of
+    		// the CompletableFuture is evaluated.
+    		try
+    		{
+    			// Stage 0: get EventHubClient for the event hub
+				retval = EventHubClient.createFromConnectionString(this.host.getEventHubConnectionString())
+				// Stage 1: use the client to get runtime info for the event hub 
+				.thenComposeAsync((ehClient) -> ehClient.getRuntimeInformation(), this.host.getExecutorService())
+				// Stage 2: extract the partition ids from the runtime info or throw on null (timeout)
+				.thenAccept((EventHubRuntimeInformation ehInfo) ->
 				{
-					this.partitionIds = ehInfo.getPartitionIds();
-	
-					TRACE_LOGGER.info(LoggingUtils.withHost(this.host.getHostName(),
-                           "Eventhub " + this.host.getEventHubPath() + " count of partitions: " + ehInfo.getPartitionCount()));
-					for (String id : this.partitionIds)
+					if (ehInfo != null)
 					{
-						TRACE_LOGGER.info(LoggingUtils.withHost(this.host.getHostName(), "Found partition with id: " + id));
+						this.partitionIds = ehInfo.getPartitionIds();
+
+						TRACE_LOGGER.info(LoggingUtils.withHost(this.host,
+				               "Eventhub " + this.host.getEventHubPath() + " count of partitions: " + ehInfo.getPartitionCount()));
+						for (String id : this.partitionIds)
+						{
+							TRACE_LOGGER.info(LoggingUtils.withHost(this.host, "Found partition with id: " + id));
+						}
 					}
-				}
-				else
+					else
+					{
+						throw new CompletionException(new TimeoutException("getRuntimeInformation returned null"));
+					}
+				})
+				// Stage 3: RUN REGARDLESS OF EXCEPTIONS -- if there was an error, wrap it in IllegalEntityException and throw
+				// Otherwise pass the partition ids along.
+				.handleAsync((empty, e) ->
```

Comment:
>handleAsync [](start = 5, length = 11)

handle will be invoked only when the previous stage (thenAccept...) failed.. was this intentional ?

Reply:
sorry my bad.. 

---
In reply to: [162722990](https://github.com/Azure/azure-event-hubs-java/pull/202#discussion_r162722990) [](ancestors = 162722990)